scico/flax/inverse.py

```diff
@@ -0,0 +1,464 @@
+            )
+            / scale,
+            s=fft_shape,
+            axes=fft_axes,
+        )
+
+        return x
+
+
+class ODPGrDescBlock(Module):
+    r"""Flax implementation of ODP gradient descent with :math:`\ell_2` loss block.
+
+    Flax implementation of the unrolled optimization with deep priors
+    (ODP) gradient descent block for inversion using :math:`\ell_2` loss
+    described in :cite:`diamond-2018-odp`.
+
+    Args:
+        operator: Operator for computing forward and adjoint mappings. In
+            this case it corresponds to the identity operator and is used
+            at the network level.
+        depth: Number of layers in block.
+        channels: Number of channels of input tensor.
+        num_filters: Number of filters in the convolutional layer of the
+            block. Corresponds to the number of channels in the output
+            tensor.
+        kernel_size: Size of the convolution filters. Default: (3, 3).
+        strides: Convolution strides. Default: (1, 1).
+        alpha_ini: Initial value of the fidelity weight `alpha`.
+            Default: 0.2.
+        dtype: Output type. Default: ``jnp.float32``.
+    """
+    operator: ModuleDef
+    depth: int
+    channels: int
+    num_filters: int
+    kernel_size: Tuple[int, int] = (3, 3)
+    strides: Tuple[int, int] = (1, 1)
+    alpha_ini: float = 0.2
+    dtype: Any = jnp.float32
+
+    def setup(self):
+        """Setting operator for batch evaluation and defining network layers."""
+        self.ah_f = lambda v: jnp.atleast_3d(
+            self.operator.adj(v.reshape(self.operator.output_shape))
+        )
+        self.a_f = lambda v: jnp.atleast_3d(self.operator(v.reshape(self.operator.input_shape)))
+
+        self.resnet = ResNet(
+            self.depth,
+            self.channels,
+            self.num_filters,
+            self.kernel_size,
+            self.strides,
+            dtype=self.dtype,
+        )
+
+        def alpha_init_wrap(rng: PRNGKey, shape: Shape, dtype: DType = self.dtype) -> Array:
+            return jnp.ones(shape, dtype) * self.alpha_ini
+
+        self.alpha = self.param("alpha", alpha_init_wrap, (1,))
+
+    def batch_op_adj(self, y: Array) -> Array:
+        """Batch application of adjoint operator."""
+        return lax.map(self.ah_f, y)
+
+    def __call__(self, x: Array, y: Array, train: bool = True) -> Array:
+        """Apply gradient descent block.
+
+        Args:
+            x: The nd-array with current stage of reconstructed signal.
+            y: The nd-array with signal to invert.
+            train: Flag to differentiate between training and testing
+                stages.
+
+        Returns:
+            The block output (i.e. next stage of inverted signal).
+        """
+
+        x = self.resnet(x, train) - self.alpha * self.batch_op_adj(lax.map(self.a_f, x) - y)
+
+        return x
+
+
+class ODPNet(Module):
+    """Flax implementation of network :cite:`diamond-2018-odp`.
```

Comment:
Perhaps an initial one-liner more consistent with the style used for the classes defined above?

Reply:
Fixed.